folding-schemes/src/folding/nova/decider_eth_circuit.rs

```diff
@@ -666,12 +780,52 @@ pub mod tests {
             GVar2,
             Pedersen<Projective>,
             Pedersen<Projective2>,
-        >::from_nova(nova);
+        >::from_nova(nova)
+        .unwrap();
 
         let cs = ConstraintSystem::<Fr>::new_ref();
 
         // generate the constraints and check that are satisfied by the inputs
         decider_circuit.generate_constraints(cs.clone()).unwrap();
         assert!(cs.is_satisfied().unwrap());
+        dbg!(cs.num_constraints());
+    }
+
+    // checks that the gadget and native implementations of the challenge computation match
+    #[test]
+    fn test_kzg_challenge_gadget() {
+        let mut rng = ark_std::test_rng();
+        let poseidon_config = poseidon_test_config::<Fr>();
+
+        let U_i = CommittedInstance::<Projective> {
+            cmE: Projective::rand(&mut rng),
+            u: Fr::rand(&mut rng),
+            cmW: Projective::rand(&mut rng),
+            x: vec![Fr::rand(&mut rng); 1],
+        };
+
+        // compute the challenge natively
+        let (challenge_W, challenge_E) =
+            KZGChallenges::<Projective>::get_challenges_native(&poseidon_config, U_i.clone())
+                .unwrap();
+
+        let cs = ConstraintSystem::<Fr>::new_ref();
+        let U_iVar =
+            CommittedInstanceVar::<Projective>::new_witness(cs.clone(), || Ok(U_i.clone()))
+                .unwrap();
+
+        let (challenge_W_Var, challenge_E_Var) =
+            KZGChallenges::<Projective>::get_challenges_gadget(
+                cs.clone(),
+                &poseidon_config,
+                U_iVar,
+            )
+            .unwrap();
+        assert!(cs.is_satisfied().unwrap());
+
+        // check that the natively computed and in-circuit computed hashes match
+        use ark_r1cs_std::R1CSVar;
```

Comment:
yes, the `ark_r1cs_std::R1CSVar` trait is needed for at the next 2 lines be able to get the `.value()` of the `challenge_W_Var, challenge_E_Var` in order to check that they match the outside the circuit values.
the import of the `use ark_r1cs_std::R1CSVar` is placed here because it is only used here and if placed as a general import it conflicts with our own `R1CSVar` struct (that checks the r1cs relation in-circuit) which is unrelated to the `ark_r1cs_std::R1CSVar` trait.